### PR TITLE
Full revamp of ihatemoney_ynh

### PR DESCRIPTION
**Pull request by Salamandar**
- [x] Use "standard" variable replacements
- [x] Use systemd
- [x] Fix the upgrade from /opt directory
- [x] implement change_url (should not be very hard)
- [x] Change some paths (socket path mostly) to allow multiple installs
- [ ] Check that the logs are correctly written

### CODE REVIEW

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
!testme (lol wtf)

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
!testme that should be good except upgrade

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
!testme should be good now !!

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
!testme

**Comment by JocelynDelalande** [comment]
Cheers, will review it for sure, thanks for that work <3. Tell me when up for review :)

**Comment by Salamandar** [comment]
Hey @JocelynDelalande you can already review the whole bunch.

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
!testme with change_url and multi_instance

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
I'm still waiting on the latest CI result, but I think this is safely mergeable. 

I still may have some changes to do on the systemd service, but i need to document myself on how Systemd handles logging.

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
Grrrr, ihatemoney has a weird behaviour when database upgrades are needed, so the systemd --line_match isn't enough.

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
@JocelynDelalande it's fully working now !

**Comment by klorydryk** [comment]
Nice work !

File: check_process
Comment (JocelynDelalande) [inline check_process:1]: ?

Comment (JocelynDelalande) [inline check_process]: ?

File: scripts/upgrade
Comment (JocelynDelalande) [inline scripts/upgrade:116]: I'd like to keep that to keep upgrade from old installations possible.

File: scripts/upgrade
Comment (JocelynDelalande) [inline scripts/upgrade:127]: same

File: scripts/upgrade
Comment (JocelynDelalande) [inline scripts/upgrade:135]: same

File: scripts/_common.sh
Comment (JocelynDelalande) [inline scripts/_common.sh]: I do not realy agree with the use of bin/activate in interactive script, makes the whole thing a big magic and statefull, and here seems to require this extra step with unset/nounset

File: scripts/upgrade
Comment (JocelynDelalande) [inline scripts/upgrade:50]: I'd like to keep that as well for not breaking upgrade paths from old versions.

File: scripts/upgrade
Comment (JocelynDelalande) [inline scripts/upgrade:66]: same

File: scripts/upgrade
Comment (JocelynDelalande) [inline scripts/upgrade:97]: same

**Review by JocelynDelalande — CHANGES_REQUESTED** [review]
I left inline comments ; and here are general remarks :

## tests

- upgrade : partly OK (see bellow)
- fresh install OK
- change_url OK (yay !)
- create project after install : KO (is that just me ?) : 
image
- backup/restore : cannot really test because of previous error
- remove : OK


## upgrade

- migrations from older versions of the package have beeen removed ; i'd like to keep previous migrations blocks that were in previous upgrade script
- there is no cleanup after upgrade (missing migration block for this upgrade) :
  - after pkg upgrade, we are left with two running daemons : one with systemd and another with supervisor
  - after pkg upgrade we are left with files in /opt, even if they are no longer used
  - same for /etc/ihatemoney


## logs

There is an issue with current MR : logs dir is deleted on remove but not created en install. Anyway, I think we could just toss away the log dir stuff : there is nothing created by the app in log dir (even with current master)  !

## conclusion


Thanks again for all that work <3 and sorry for the delay.
I hope we will make it to master soon :). Are you OK to address the points I highlighted ?

**Comment by Salamandar** [comment]
Ah yes, this "CSRF token" error... I've had that with the current package some months ago. Going into private browsing fixed the issue, but clearing cache/cookies did not help.

That's clearly not an issue with my pull request, but with ihatemoney itself (or maybe the nginx config ? idk.).

File: check_process
Comment (Salamandar) [reply check_process]: Hmmm I may have "rebased" on the example_ynh, i don't recall exactly why i edited this but I'll edit it back.

File: scripts/upgrade
Comment (Salamandar) [reply scripts/upgrade:50]: It's useless as the install directory completely changed.

**Comment by Salamandar** [comment]
@JocelynDelalande The whole venv-old backup style is not needed anymore, since Yunohost already provides a backup/restore system (and the app is backed up before upgrading, and restored on failure).

The whole venv update though, i'm working on it.

**Comment by Salamandar** [comment]
!testme
(yeah, went from virtualenv to venv, checking is required)

File: scripts/_common.sh
Comment (Salamandar) [reply scripts/_common.sh]: What's the problem with activate ? Here it's encapsulated in a subshell.

**Comment by alexAubin** [comment]
```
21676 INFO DEBUG - + eval 'pip install --upgrade pip setuptools>=18.5 gunicorn>=19.3.0 PyMySQL>=0.9,<0.10 SQLAlchemy<1.4 ihatemoney>=4,<5'
21677 INFO DEBUG - ++ pip install --upgrade pip setuptools gunicorn PyMySQL SQLAlchemy ihatemoney
21677 INFO WARNING - _common.sh: line 38: 0.10: No such file or directory
```

Lolz it looks like it interprets the `>=` and `<0.10` as file redirects...

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
!testme

**Comment by Salamandar** [comment]
@JocelynDelalande It's finally working and all your comments have been addressed :)

File: scripts/_common.sh
Comment (Salamandar) [reply scripts/_common.sh]: I finally removed the activate import.

File: doc/DISCLAIMER.md
Comment (alexAubin) [inline doc/DISCLAIMER.md]:
```suggestion
* LDAP authentication and Single Sign-on is not supported. The login mechanism in IHateMoney is per-project (not per-user) and therefore can't be integrated in YunoHost.

```

File: doc/DISCLAIMER_fr.md
Comment (alexAubin) [inline doc/DISCLAIMER_fr.md]:
```suggestion
* L'authentification LDAP et login unifié (SSO) n'est pas supportée. Le mécanisme de connexion sur IHateMoney se fait par projet et ne peut donc pas être intégrée dans YunoHost

```

File: manifest.json
Comment (alexAubin) [inline manifest.json]:
```suggestion
        "en": "Simple app to manage your collective expenses",
        "fr": "Gérez simplement vos dépenses collectives"
```

File: manifest.json
Comment (alexAubin) [inline manifest.json]:
```suggestion
    "upstream": {
        "license": "MIT",
        "website": "https://github.com/spiral-project/ihatemoney",
        "admindoc": "https://ihatemoney.readthedocs.org/",
        "code": "https://github.com/spiral-project/ihatemoney",
        "demo": "https://ihatemoney.org/demo/"
    },
```

File: conf/ihatemoney.cfg
Comment (alexAubin) [inline conf/ihatemoney.cfg:7]: Sounds like a good idea, I'm wondering if we shouldn't make this a default behavior in yunohost core (c.f. similar situation in nginx conf)

**Comment by JocelynDelalande** [comment]
@Salamandar you rocks <3 . Reviewing.

btw, I do agree with all @alexAubin suggestions, and you ? I let you approve them if so.

**Comment by JocelynDelalande** [comment]
> venv-old



> @JocelynDelalande The whole venv-old backup style is not needed anymore, since Yunohost already provides a backup/restore system (and the app is backed up before upgrading, and restored on failure).

fair enough.

File: scripts/_common.sh
Comment (JocelynDelalande) [inline scripts/_common.sh:38]:
```suggestion
    "$venv_dir/bin/pip install --upgrade pip "${pip_packages[@]}"
```

File: scripts/restore
Comment (JocelynDelalande) [inline scripts/restore]:
```suggestion
ynh_script_progression --message="Reloading NGINX web server..." --weight=1
```

File: scripts/upgrade
Comment (JocelynDelalande) [inline scripts/upgrade]: (pipenv ?)

File: scripts/upgrade
Comment (JocelynDelalande) [inline scripts/upgrade:58]: Better not loose that please.

File: scripts/upgrade
Comment (JocelynDelalande) [inline scripts/upgrade:38]: nice

File: scripts/upgrade
Comment (JocelynDelalande) [inline scripts/upgrade:136]: 😍 I did'nt know it, this is so helpfull !

File: scripts/upgrade
Comment (JocelynDelalande) [inline scripts/upgrade:103]: I'd like to keep that upgrade migration also.

File: scripts/upgrade
Comment (JocelynDelalande) [inline scripts/upgrade:128]: same

File: conf/ihatemoney.cfg
Comment (JocelynDelalande) [inline conf/ihatemoney.cfg:7]:
```suggestion
__SUB_PATH_ONLY__ APPLICATION_ROOT='__PATH__'
```
missing space made me think that was a full variable name

File: scripts/upgrade
Comment (JocelynDelalande) [inline scripts/upgrade:164]:
```suggestion
# Allows to comment some config lines if not using sub path
sub_path_only="$(if [[ "$path_url" == "/" ]]; then echo '# ' ; else echo ''; fi)"
```

(btw, nice trick <3)

**Review by JocelynDelalande — CHANGES_REQUESTED** [review]
That is realy nice :).

I nitpicked a few easy stuff/suggestions in comment/strings mainly

The only important thing to me is keeping upgrade migrations, I highlighted 3 of them that I would like to keep. 

I'd like to declare them using explicit version check as you did for opt stuff. I can commit the required edits if you prefer tell me.

And I learnt interesting stuff reading your code, thanks a lot for that also :).

**Comment by Salamandar** [comment]
> btw, I do agree with all @alexAubin suggestions, and you ? I let you approve them if so.

Yes, I'm on it :)

File: manifest.json
Comment (Salamandar) [reply manifest.json]: Was just the demo added ? The github diff is weird here.

Comment (alexAubin) [reply manifest.json]: Ah yes sorry indeed, just adding the demo

File: scripts/upgrade
Comment (alexAubin) [reply scripts/upgrade:164]: Btw can be further improved with:

```suggestion
sub_path_only="$([[ "$path_url" == "/" ]] && echo '# ' || echo '')"
```

(but maybe more cryptic / too compact depending on how familiar you are with bash ;))

File: scripts/_common.sh
Comment (Salamandar) [reply scripts/_common.sh:38]: Well actually I prefer your suggestion, but it is deprecated, or dangerous :
https://www.py4u.net/discuss/185791

I don't understand if it's especially for "install pip" or if it's for every pip commands though…

File: scripts/upgrade
Comment (Salamandar) [reply scripts/upgrade:164]: @alexAubin that's not correct. Here's an example:
```
$ if true; then false; else echo ok; fi
$ true && false || echo ok
ok
```
With &&/||, if the second command fails (the if-then block), then the third is executed (the if-else block).
That's not the case with if-then-else.


EDIT: Obviously the echo won't fail, but that syntax is dangerous enough that I prefer not giving the habit of using it to people who aren't aware of that.

File: conf/ihatemoney.cfg
Comment (Salamandar) [reply conf/ihatemoney.cfg:7]: That should be tested.

The cfg file is a python script. Here will be the corresponding code :
```
blabla
 APPLICATION_ROOT='the_path'
blabla
```

See the space at the start of the line ? This will trigger a `IndentationError: unexpected indent`.

Comment (JocelynDelalande) [reply conf/ihatemoney.cfg:7]: you're right. So long for readability :)

File: scripts/upgrade
Comment (alexAubin) [reply scripts/upgrade:164]: Fair enough ;)

File: manifest.json
Comment (Salamandar) [reply manifest.json]: Ok ! Merging.

File: scripts/upgrade
Comment (Salamandar) [reply scripts/upgrade]: Fixed ! I temporarily tested pipenv before going back to venv.

venv is the official tool, but people say pipenv is better. Can't take a decision about it… I may try pipenv for a second pull request.

File: scripts/upgrade
Comment (Salamandar) [reply scripts/upgrade:58]: So, does that just invalidate session cookies ?
If yes, that would be even better to do that at every upgrade, no ?

File: scripts/upgrade
Comment (Salamandar) [reply scripts/upgrade:103]: That's basically already done in the "if $upgrade_from_opt" block lines ~85.

Comment (Salamandar) [reply scripts/upgrade:103]: Also, as i'm now using systemd, i don't configure it anymore.
Though, I should probably remove its config.

Comment (Salamandar) [reply scripts/upgrade:103]: Done. The nginx config is correctly updated, and the supervisor config is delete.

File: scripts/upgrade
Comment (Salamandar) [reply scripts/upgrade:128]: The INSTALL_DIR is fully deleted (line ~87) so that's included in that.

File: scripts/upgrade
Comment (Salamandar) [reply scripts/upgrade:58]: Done. Now, at every change (install, change_url, update) the secret_key is generated and it's not saved in the settings anymore.

**Comment by Salamandar** [comment]
!testme

File: scripts/upgrade
Comment (alexAubin) [reply scripts/upgrade]: (I never understood why people complain about regular venv ... Imho regular venv are just fine ... And imho they make much more sense than the madness happening in other languages where there's either no venv (php? except when using composer maybe... or nodejs with their crazy node module and you never know if you should be using yarn or npm or nvm or whatever))

**Comment by Salamandar** [comment]
!testme

**Comment by JocelynDelalande** [comment]
@Salamandar current status in my brain : waiting for you to fix upgrade and change_url to merge. OK for you ?

(Maybe ruling the obvious, but I want to be sure we are not waiting for each other pointlessly :p)

(once that will be merged, I'll upgrade pkg to latest ihatemoney version that was released couple of days ago #42 )

**Comment by Salamandar** [comment]
@JocelynDelalande yeah don't worry.
Also, I would need some help from @alexAubin about the failure during the upgrade. Can't understand the log AT ALL.

File: scripts/upgrade
Comment (JocelynDelalande) [inline scripts/upgrade]:
```suggestion
"$final_path/venv/bin/python3" -m pip install --upgrade pip "${pip_dependencies[@]}"
```

(otherwise, setuptools is mentioned twice and pip issues a warning)

Comment (JocelynDelalande) [reply scripts/upgrade]: @Salamandar suprisingly, applying that patch made the difference for me (I guess pip returns nonzero when issuing this warning) and made the upgrade pass !

Comment (Salamandar) [reply scripts/upgrade]: Weird, but thanks for the info ! I guess setuptools isn't needed at all…

Comment (Salamandar) [reply scripts/upgrade]: Ah, my bad, it's already in the pip_dependencies. I guess there's a corner case when setuptools is specified twice…
Still, that's weird that it fails in the upgrade and not in the new install, and not when I try this code manually…

**Comment by Salamandar** [comment]
!testme for the upgrade !

**Comment by Salamandar** [comment]
!testme

Oops @JocelynDelalande I re-requested review, missclick.

**Comment by Salamandar** [comment]
!testme

I removed the "backup before upgrade" to workaround the yunohost log issue.

**Comment by Salamandar** [comment]
!testme

Should be good now !!

**Comment by Salamandar** [comment]
FINALLY ! @JocelynDelalande it's working !

**Comment by alexAubin** [comment]
merge

**Comment by brimwats1** [comment]
this might not have been right, but I really wanted to try IHM, and it failed when I installed so I tried installing this testing branch and it also failed (error 503). Probably more likely I did something wrong but I thought I'd offer my installation and removal logs

https://paste.yunohost.org/raw/ziholazone


https://paste.yunohost.org/raw/etabobajow

**Comment by brimwats1** [comment]
looks like it's still pending @JocelynDelalande :)

**Comment by ericgaspar** [comment]
hey! shouldn't we merge this?

Edit: I get this error at install  `yunohost: error: unrecognized arguments: --log_type=systemd`

File: scripts/install
Comment (ericgaspar) [inline scripts/install]:
```suggestion
yunohost service add $app --description="$app daemon for IHateMoney" --log=systemd
```

File: scripts/restore
Comment (ericgaspar) [inline scripts/restore]:
```suggestion
yunohost service add $app --description="$app daemon for IHateMoney" --log=systemd
```

File: scripts/upgrade
Comment (ericgaspar) [inline scripts/upgrade]:
```suggestion
yunohost service add $app --description="$app daemon for IHateMoney" --log=systemd
```

**Review by ericgaspar — CHANGES_REQUESTED** [review]
fix: unrecognized arguments: --log_type

**Comment by Salamandar** [comment]
@ericgaspar is this a new change ? I didn't have those errors.

**Comment by Salamandar** [comment]
@ericgaspar Still waiting on @JocelynDelalande . Is he OK ? 😞

**Review by ericgaspar — APPROVED** [review]

**Comment by ericgaspar** [comment]
Let's try to push this PR forward as the master branch is failing tests

**Pull request merged**
